### PR TITLE
embed resolvewithplus

### PR DESCRIPTION
**Pull request by iambumblehead**
This PR embeds resolvewithplus inside the esmock package. To test the embedded result in windows-latest ci eval'd node scripts were used rather than `sed` and `cp` commands.

### CODE REVIEW

**Comment by iambumblehead** [comment]
@koshic @aladdin-add @tripodsan please review

related discussions here https://github.com/iambumblehead/esmock/pull/179#discussion_r999320947 and here https://github.com/iambumblehead/esmock/issues/176#issuecomment-1281549228

**Comment by iambumblehead** [comment]
It might be better if we _do not enable_ import.meta.resolve, since resolvewithplus will be embedded and is already being used. Not sure.

**Comment by iambumblehead** [comment]
the comment here indicates `import.meta.resolve` is not reliable https://github.com/nodejs/node/issues/30682#issuecomment-950047088 and based on discussions I've read, it hasn't been updated at nodejs upstream in a long time maybe a few years.

File: tests/package.json [hunk 1]
```diff
@@ -27,6 +27,9 @@
     "mini:default": "cd .. && npx esbuild ./src/*js --minify --allow-overwrite --outdir=src",
     "mini:win32": "cd .. && cd src && forfiles /m \"*.js\" /c \"cmd /c npx esbuild @file --minify --allow-overwrite --outfile=@file\"",
     "mini": "run-script-os",
+    "embed:cp-refd": "node --input-type=module -e \"import fs from 'fs';fs.copyFileSync('../node_modules/resolvewithplus/resolvewithplus.js', '../src/resolvewithplus.js')\"",
+    "embed:update-refs": "node --input-type=module -e \"import fs from 'fs';let p='../src/esmockModule.js';fs.writeFileSync(p, fs.readFileSync(p, 'utf-8').replace(/'resolvewithplus'/, '\\'./resolvewithplus.js\\''))\"",
+    "embed": "npm run embed:update-refs && npm run embed:cp-refd",
```
Comment (aladdin-add) [inline tests/package.json:32]: not sure I've gotten your point: `npm run embed` should be run before publish, but I didn't see it was added.

I had a concern: the code in here is really fallible, and hard to debug. can we just copy the `resolvewithplus` to a folder, and provide a script to update it to the newer version when you want?

I made a similar thing in `eslint-plugin-n`:
* https://github.com/eslint-community/eslint-plugin-n/blob/master/scripts/convert-pure-esm-to-cjs.js
* https://github.com/eslint-community/eslint-plugin-n/blob/master/lib/converted-esm/import-meta-resolve.js

**Review by aladdin-add — CHANGES_REQUESTED** [review]

File: tests/package.json (hunk 1 above)
Comment (iambumblehead) [reply tests/package.json:32]: `npm run embed` is used by `npm run test:all-ci`, also used by the prepublish script,

combining parts of the main and test package.json scripts,
```json
{
    "embed:cp-refd": "node --input-type=module -e \"import fs from 'fs';fs.copyFileSync('../node_modules/resolvewithplus/resolvewithplus.js', '../src/resolvewithplus.js')\"",
    "embed:update-refs": "node --input-type=module -e \"import fs from 'fs';let p='../src/esmockModule.js';fs.writeFileSync(p, fs.readFileSync(p, 'utf-8').replace(/'resolvewithplus'/, '\\'./resolvewithplus.js\\''))\"",
    "embed": "npm run embed:update-refs && npm run embed:cp-refd",
    "mini:pkg": "npm pkg delete scripts devDependencies",
    "test:all-ci": "npm run mini && npm run embed && npm run test:all",
    "test-ci": "npm run test:install && npm run test:all-ci",
    "prepublishOnly": "npm run lint && npm run test-ci && npm run mini:pkg"
}
```
If storing resolvewithplus directly inside the src folder is preferred that is OK with me. The above are complicated and ugly, but they are verified at the unit-test pipeline.


[0]: https://github.com/iambumblehead/esmock/blob/master/package.json#L79

Comment (iambumblehead) [reply tests/package.json:32]: a little ugly but not too terrible (for me)

_embed:cp-refd_
```javascript
import fs from 'fs';
fs.copyFileSync('../node_modules/resolvewithplus/resolvewithplus.js', '../src/resolvewithplus.js')
```
_embed:update-refs_
```javascript
import fs from 'fs';
let p='../src/esmockModule.js';
fs.writeFileSync(p, fs.readFileSync(p, 'utf-8').replace(/'resolvewithplus'/, '\'./resolvewithplus.js\''))

Comment (aladdin-add) [reply tests/package.json:32]: I slightly prefer storing in the src, to avoid the magic like `embed:cp-refd`/`embed:update-refs`, but let's see what others think. :)

Comment (iambumblehead) [reply tests/package.json:32]: if one trusts these lines, they allow one to develop esmock without needing to manually sync resolvewithplus and without needing to add extra files. esmock can be developed and used with regular npm, while unit-tests and prepublishOnly automaticaly get the embedded result.

Comment (iambumblehead) [reply tests/package.json:32]: my internet connection is slow I did not see your reply when adding my last message

Comment (iambumblehead) [reply tests/package.json:32]: @koshic @tripodsan what are your opinions? :)

Comment (iambumblehead) [reply tests/package.json:32]: Using current approach, it remains easy to reference [new resolvewithplus branches.][0] Simply update package.json and use npm as usual. No need to sync a file or include sync file changes in the PR,
```json
{
  "dependencies": {
    "resolvewithplus": "iambumblehead/resolvewithplus#return-encoded-url-same-as-import.meta.resolve"
  }
}
```
The other approach, committing resolvewithplus.js into the repo, creates work and adds noise to the PR, where sync file changes become an added part of the PR.

[0]: https://github.com/iambumblehead/resolvewithplus/pull/40

**Comment by koshic** [comment]
> It might be better if we _do not enable_ `import.meta.resolve`, since resolvewithplus will be embedded and is already being used. Not sure. Tests pass when `import.meta.resolve` is used so maybe its ok.

Why don't use 'readfilewithplus'? Every Node function / API can be replaced, the question in only about reasons. Personally for me, they are:

1. Functionality is outdated and work with errors / extremely slow , Node team don't ~~want~~ have enough time to fix ~~it due to working on very important test runner~~. You have to choose - to wait or to work.
2. Functionality is poor (comparing to the rest of the world), like missed fs.rm which caused a lot of packages like rimraf, del, etc. 
3. Functionality is ok, but some your requirements are out of scope (intercepting / configurability / .xxx extension support / doesn't matter). This is the root cause for 'resolve hell', when every bundler / compiler / etc. have to write CommonJS (at the first time) own resolver instead of use existing 'require.resolve'.
4. Functionality doesn't exist. Just no way to do what you want. Like early ESM which caused `@std/esm`, 'node --test', 'node --watch', etc.
5. Functionality is experimental. Tricky moment, due to some features less experimental than others and can be used as stable things. Common example - es6 (and other TC39 proposals), which used a long time with transpilers only. Yes, landed implementation have some differences. Yes, it's not native. Many 'yes' and one 'but it looks nice, let's use!!!".

Speaking of meta resolve, it's 5 until we have proves for 1 or do not have additional requirements like dynamically changing [conditions](https://nodejs.org/dist/latest-v19.x/docs/api/packages.html#conditional-exports) to be 3. 

In other words, you can consider this as 'new functionality and readfilewithplus as polyfill', which is sooner or later will be enabled by default. If you don't have obligations like 'code must not require maintenance in 10 years', signed in blood )

**Comment by koshic** [comment]
> the comment here indicates `import.meta.resolve` is not reliable [nodejs/node#30682 (comment)](https://github.com/nodejs/node/issues/30682#issuecomment-950047088) and based on discussions I've read, it hasn't been updated at nodejs upstream in a long time maybe a few years.

It's a kind of not true. Meta resolve supports [imports](https://nodejs.org/dist/latest-v19.x/docs/api/packages.html#subpath-imports), while Jest (and a lot of custom resolvers) doesn't support it, and TS gets such support a few months ago. 

The main advantage of native function - the highest probability that result will be aligned with other native functionality, and any deviations will be detected by community and fixed (or be well-known). 

At the page you provided I can t find any issues with import.meta.resolve. May be it doesn't work with workers? May be there are very special setup? May be someone just haven't enough developer skill? I don't know, really )

File: tests/package.json (hunk 1 above)
Comment (koshic) [reply tests/package.json:32]: The main question - what is our goal in terms of package internal structure / development workflow / etc. What do I mean? 

Build / compile / bundle strep is not something new, and my personal preference (due to a multiple factors) is to use resolvewithplus package as devDependency and just call 'build' before publishing. 'Build' can contains absolutely anything, from 'cp' to 'webpack'. But esbuild is better ). This pattern is transparent and not introduce any kind of magic.

BUT if you want to avoid external tools usage, specially designed tests or even have compilerphobia, it may make sense. And 'cp' pattern is not so ugly for me. You know, @aladdin-add knows, I know - that's enough.

PS Sure, in my current company developer will be forced to rewrite 'that crap' (especially if somebody adds me to the review for old times sake), just because he is not alone and nobody wants to understand his magic in the future.

Comment (iambumblehead) [reply tests/package.json:32]: > Build / compile / bundle strep is not something new, and my personal preference (due to a multiple factors) is to use resolvewithplus package as devDependency and just call 'build' before publishing.

What do you think of this idea @aladdin-add @koshic?
  1. move resolvewithplus to devDependency, per @koshic
  2. update esbuild commands to output single bundle file, per @koshic 
  3. esbuild output saved to src/ with the package-exported name 'esmockLoader.js'. This way, esmock does not need a 'dist' or 'build' folder and can continue to be cloned from git and used as-is without build step,
  4. other files are deleted

final output would look like this,
```console
src/
├── esmock.d.ts
└── esmockLoader.js
```


[0]: https://github.com/iambumblehead/esmock/blob/master/tests/package.json#L27

Comment (iambumblehead) [reply tests/package.json:32]: esbuild is rigid and may not allow us to output a file the same name, which brings us back to the problem of missing rm and cp commands in windows-latest pipelines

Comment (koshic) [reply tests/package.json:32]: Let me play with master & esbuild, it will take ~30m

Comment (koshic) [reply tests/package.json:32]: Place the following code in build.js into the package root:

```js
import { fileURLToPath } from "node:url";
import { build } from "esbuild";

await build({
  outdir: ".",
  entryPoints: ["esmockModule.js"],
  bundle: true,
  // can be globbed with any glob package like ['src/*', '!/src/esmockModule.js']
  external: ["./esmockErr.js", "./esmockCache.js"],
  // can be ommited via 'cd src' before
  absWorkingDir: fileURLToPath(new URL("src", import.meta.url)),
  format: "esm",
  platform: "node",
  // to be sure that nothing was changed, very safe to remove
  target: "node18",
  allowOverwrite: true,
});

```

It can be simplified via running it from 'src' dir, or called via command line with the same options.

Comment (iambumblehead) [reply tests/package.json:32]: I'll be un-available to respond until later this evening. Thanks for the discussion and script demo

Comment (koshic) [reply tests/package.json:32]: I'm confused a bit )

My proposal (and code) is about adding one step right before package publishing - to bundle 'resolvewith' into esmockModule.js. There is no reason to run bundle step for tests, 'resolvewith' should stay as external dependency and they should work with original esmockModule.js.

Comment (iambumblehead) [reply tests/package.json:32]: The final, bundled, minified version is unit-tested to ensure published sources pass all tests, before they are published. If tests fail, the publish step is aborted.

Comment (koshic) [reply tests/package.json:32]: Thx, now I catch it   we misunderstood each other )

I proposed 'precision strike' - modify only one file (esmockModule.js) during 'bundle' step to keep your current flow. That step can be executed before tests (and followed publish), as you prefer.

But you decided to solve this task more radically - to bundle all .js files into one, then test & publish resulting bundle (esmockLoader.js) plus types. Right?

Comment (iambumblehead) [reply tests/package.json:32]: yes :) https://github.com/iambumblehead/esmock/pull/181#discussion_r1000842074
> update esbuild commands to output single bundle file, per @koshic

final output would look like this,
```console
src/
├── esmock.d.ts
└── esmockLoader.js
```

Comment (koshic) [reply tests/package.json:32]: Nice ) Only one note: you don't need 'src' anymore in the final structure. Types can be moved the at root without any concerns, bundle - with minor tweaks. What do you think? As I see, you really like minimal style )

Comment (iambumblehead) [reply tests/package.json:32]: If we generate output at src, we can continue pointing [exports.import][0] at the src location. The benefit: the project can be cloned and used right away without build step.

It would be good to export esmock rather than esmockLoader maybe in a follow up PR.


[0]: https://github.com/iambumblehead/esmock/blob/master/package.json#L10-L13

Comment (iambumblehead) [reply tests/package.json:32]: I'll update the PR again in a day or two

Comment (koshic) [reply tests/package.json:32]: > If we generate output at src, we can continue pointing [exports.import](https://github.com/iambumblehead/esmock/blob/master/package.json#L10-L13) at the src location. The benefit: the project can be cloned and used right away without build step.

Note for the next PR: publishConfig can keep this benefit and provide different exports for different situations.

Comment (iambumblehead) [reply tests/package.json]: @koshic @aladdin-add unusual "ref" scripts are removed and esbuild "bundle" is used. What do you think?
```console
npx esbuild esmockLoader.js \
  --minify \
  --bundle \
  --allow-overwrite \
  --platform=node \
  --format=esm \
  --outfile=esmockLoader.js
```

Comment (iambumblehead) [reply tests/package.json]: thank you for sparing me your time so far

**Review by tripodsan — APPROVED** [review]

File: tests/package.json (hunk 1 above)
Comment (koshic) [reply tests/package.json]: I'd prefer two entry different points in package.json - one for '--loader', one for 'import esmock' inside test code. Why?

To avoid mixing code executed in different sandboxes for different purposes, and possible side-effects.

So, ie may look like '--loader esmock/loader' and 'import {xxx} from "esmock"' from regular code:
```
exports: {
  "types": "...",
  "import": {
    ".": "./dist/main.js",
    "./loader": "./dist/loader.js",
  }
}
```

There is one thing - the more the build result differs from the source files, the less sense it makes to test this result. As an example, for C++ code you will never write tests to check .exe file, right?

Here is the same: bundle(s) is not the best source for unit tests (that's why I use 'dist' dir in my example). It can be tested in kind of e2e tests ('loaded without errors', 'has exports A & B', etc.), but regular unit tests is about raw source by many reasons: you can have multiple output formats, project can migrate to TS, you may want to change bundler (and output) without touching unit tests, and so on.

So, I propose to have independent 'source with unit tests' & 'bundled code with e2e tests' layers. And it a good point for discussion - but in a different PR. This is PR is ready to merge.

**Review by koshic — APPROVED** [review]

File: tests/package.json (hunk 1 above)
Comment (iambumblehead) [reply tests/package.json]: > I'd prefer two different entry points in package.json - one for '--loader', one for 'import esmock' inside test code. Why?

The current interface using `--loader=esmock` and`import esmock from ''esmock"` is easier for users and documentation. 


> To avoid mixing code executed in different sandboxes for different purposes, and possible side-effects.

The loader will [run in a separate thread][0] soon and all tests currently pass for supported versions of node.


> There is one thing - the more the build result differs from the source files, the less sense it makes to test this result.

The un-processed sources are tested at the [code coverage pipeline,][1] but only on the node18 image. This PR (and other PRs) did and do encounter bugs related to bundling and minification and it is useful to test these files before publishing them.


> As an example, for C++ code you will never write tests to check .exe file, right?

This is true generally, but there is a unique context for esmock. Functional tests give coverage that is nearly 100% and exporting and unit-testing each function is not needed imho. Input and output results are predictable.



[0]: https://github.com/nodejs/node/pull/44710
[1]: https://github.com/iambumblehead/esmock/actions/runs/3303674831/jobs/5451893808#step:7:558

Comment (iambumblehead) [reply tests/package.json]: @aladdin-add this PR waits or you approval :)

**Review by aladdin-add — APPROVED** [review]
LGTM!
